### PR TITLE
[mypy] [9891] Reduce mypy errors

### PR DESCRIPTION
**Pull request by rodrigc**
https://twistedmatrix.com/trac/ticket/9891

Clean up mypy errors such as:

```
+src/twisted/internet/tcp.py:1027:5: error: Signature of "_FileDescriptorReservation" incompatible with "__exit__" of supertype "_IFileDescriptorReservation"  [override]
+        def __exit__(self, excValue, excType, traceback):
+        ^
+src/twisted/internet/tcp.py:1069:5: error: Signature of "_NullFileDescriptorReservation" incompatible with "__exit__" of supertype "_IFileDescriptorReservation"
+[override]
+        def __exit__(self, excValue, excType, traceback):
+        ^
+src/twisted/internet/testing.py:871:5: error: Signature of "RaisingMemoryReactor" incompatible with "adoptDatagramPort" of supertype "IReactorSocket"  [override]
+        def adoptDatagramPort(self, fileDescriptor, addressFamily, protocol,
+        ^
+src/twisted/internet/testing.py:880:5: error: Signature of "RaisingMemoryReactor" incompatible with "adoptStreamConnection" of supertype "IReactorSocket"  [override]
+        def adoptStreamConnection(self):
+        ^
+src/twisted/test/stdio_test_halfclose.py:58:5: error: Signature of "HalfCloseProtocol" incompatible with "writeConnectionLost" of supertype "IHalfCloseableProtocol"
+[override]
+        def writeConnectionLost(self, reason):
+        ^
+src/twisted/conch/scripts/cftp.py:40:21: error: Incompatible types in assignment (expression has type "List[List[Union[str, int, None]]]", base class "ConchOptions"
+defined the type as "List[List[Optional[str]]]")  [assignment]
+        optParameters = [
+                        ^
+src/twisted/web/_auth/wrapper.py:243:5: error: Signature of "HTTPAuthSessionWrapper" incompatible with "putChild" of supertype "IResource"  [override]
+        def putChild(self, child):
+        ^
+src/twisted/python/test/test_shellcomp.py:485:31: error: List item 1 has incompatible type "None"; expected "str"  [list-item]
+        optFlags = [['no-stalls', None,
+                                  ^
+src/twisted/mail/test/test_smtp.py:434:22: error: Argument 1 to "append" of "list" has incompatible type
+"Tuple[bytes, bytes, bytes, Tuple[bytes, Tuple[bytes, bytes, List[bytes], bytes]]]"; expected "Tuple[bytes, bytes, None, None]"  [arg-type]
+            data.append((b'DATA\r\n', b'354.*\r\n',
+                         ^
+src/twisted/conch/test/test_userauth.py:246:5: error: Signature of "AnonymousChecker" incompatible with "requestAvatarId" of supertype "ICredentialsChecker"  [override]
+        def requestAvatarId(credentials):
+        ^
+src/twisted/internet/test/test_tcp.py:462:5: error: Signature of "FakeResolver" incompatible with "getHostByName" of supertype "IResolverSimple"  [override]
+        def getHostByName(self, name, timeout):
+        ^
```

### CODE REVIEW

File: src/twisted/internet/tcp.py [hunk 1]
```diff
@@ -949,24 +949,6 @@ def reserve():
...
 
 
-    def __enter__():
-        """
-        Release the underlying file descriptor so that code within the
-        context manager can open a new file.
-        """
-
-
-    def __exit__(excType, excValue, traceback):
```
Comment (adiroiban) [inline src/twisted/internet/tcp.py]: I think that this should not be removed. The enter and exit method are required for the context manager behaviour.

```
Interface      -> def __exit__(      excType, excValue, traceback):
Implementation -> def __exit__(self, excValue, excType, traceback):
```

The interface is correct... the order is `type, value, traceback`.
The implementation should be fixed.

Here is the documentation https://docs.python.org/3/reference/datamodel.html#object.__enter__

```python
object.__exit__(self, exc_type, exc_value, traceback)
```

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks for the changes. It looks good.
There is a minor blocker for this.

Please fix and resubmit.

Thanks again!

File: src/twisted/internet/tcp.py (hunk 1 above)
Comment (rodrigc) [reply src/twisted/internet/tcp.py]: That didn't work.  I changed to:

```
diff --git a/src/twisted/internet/tcp.py b/src/twisted/internet/tcp.py
index 0f493c091..e8f955317 100644
--- a/src/twisted/internet/tcp.py
+++ b/src/twisted/internet/tcp.py
@@ -956,7 +956,7 @@ class _IFileDescriptorReservation(Interface):
         """


-    def __exit__(excType, excValue, traceback):
+    def __exit__(exc_type, exc_value, traceback):
         """
         Attempt to re-open the reserved file descriptor.  See
         L{reserve} for caveats.
@@ -1024,7 +1024,7 @@ class _FileDescriptorReservation(object):
         self._fileDescriptor = None


-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, exc_value, exc_type, traceback):
         """
         See L{_IFileDescriptorReservation.__exit__}.
         """
@@ -1066,7 +1066,7 @@ class _NullFileDescriptorReservation(object):
         """


-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, exc_value, exc_type, traceback):
         """
         Do nothing.  See L{_IFileDescriptorReservation.__exit__}.
```

and still got:

```
src/twisted/internet/tcp.py:1027:5: error: Signature of "_FileDescriptorReservation" incompatible with "__exit__" of supertype "_IFileDescriptorReservation"  [override]
src/twisted/internet/tcp.py:1069:5: error: Signature of "_NullFileDescriptorReservation" incompatible with "__exit__" of supertype "_IFileDescriptorReservation"  [override]
```

Comment (adiroiban) [reply src/twisted/internet/tcp.py]: I have applied the following patch on top of `f541fa566 Craig Rodrigues 2020-07-02 Remove __enter__() and __exit__() from  _IFileDescriptorReservation
` and it worked

```diff
diff --git a/src/twisted/internet/tcp.py b/src/twisted/internet/tcp.py
index 79c049a1a..b69119a39 100644
--- a/src/twisted/internet/tcp.py
+++ b/src/twisted/internet/tcp.py
@@ -1006,7 +1006,7 @@ class _FileDescriptorReservation(object):
         self._fileDescriptor = None
 
 
-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, excType, excValue, traceback):
         """
         See L{_IFileDescriptorReservation.__exit__}.
         """
```

Comment (rodrigc) [reply src/twisted/internet/tcp.py]: Oh OK, this worked:

```
--- a/src/twisted/internet/tcp.py
+++ b/src/twisted/internet/tcp.py
@@ -1024,7 +1024,7 @@ class _FileDescriptorReservation(object):
         self._fileDescriptor = None


-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, excType, excValue, traceback):
         """
         See L{_IFileDescriptorReservation.__exit__}.
         """
@@ -1066,7 +1066,7 @@ class _NullFileDescriptorReservation(object):
         """


-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, excType, excValue, traceback):
         """
         Do nothing.  See L{_IFileDescriptorReservation.__exit__}.
```

Comment (rodrigc) [reply src/twisted/internet/tcp.py]: Fixed

**Review by adiroiban — APPROVED** [review]
All good. Thanks!

**Pull request merged**